consensus/transactions.go

```diff
@@ -164,19 +165,27 @@ func (s *State) applyTransaction(t Transaction) (outputDiffs []OutputDiff, contr
 // match the sum of all the inputs. Outputs created by storage proofs are not
 // considered, as they were already considered when the contract was created.
 func (t Transaction) OutputSum() (sum Currency) {
+	// NOTE: manual overflow checking is performed here to prevent redundant
+	// checks.
+
 	// Add the miner fees.
 	for _, fee := range t.MinerFees {
-		sum += fee
+		sum.Add(fee)
 	}
 
 	// Add the contract payouts
 	for _, contract := range t.FileContracts {
-		sum += contract.Payout
+		sum.Add(contract.Payout)
 	}
 
 	// Add the outputs
 	for _, output := range t.Outputs {
-		sum += output.Value
+		sum.Add(output.Value)
+	}
+
+	// Check for overflow
+	if sum.Overflow() {
+		// TODO: ???
```

Comment:
In a formal spec I would say its behavior is undefined, but as it's written now it will behave as you said, as though the overflow did not occur. That is, `Currency(2^128 + 9001).Cmp(2^128) == 1`.